src/lib/scheduler/operator_task.cpp

```diff
@@ -1,59 +1,84 @@
-    const std::shared_ptr<AbstractOperator>& op, std::vector<std::shared_ptr<AbstractTask>>& tasks,
-    std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<AbstractTask>>& task_by_op) {
-  const auto task_by_op_it = task_by_op.find(op);
-  if (task_by_op_it != task_by_op.end()) return task_by_op_it->second;
+using namespace opossum;  // NOLINT
 
-  auto task = std::make_shared<OperatorTask>(op);
-  task_by_op.emplace(op, task);
+/**
+ * Create tasks recursively. Called by `make_tasks_from_operator`.
+ * @returns the root of the subtree that was added.
+ * @param task_by_op is used as a cache to avoid creating duplicate tasks. For example, in the case of diamond shapes.
+ */
+std::shared_ptr<OperatorTask> add_operator_tasks_recursively(const std::shared_ptr<AbstractOperator>& op,
+                                                             std::unordered_set<std::shared_ptr<OperatorTask>>& tasks) {
+  auto task = op->get_or_create_operator_task();
+  // By using an unordered set, we can avoid adding duplicate tasks.
+  const auto& [_, inserted] = tasks.insert(task);
+  if (!inserted) return task;
 
   if (auto left = op->mutable_left_input()) {
-    if (auto left_subtree_root = _add_tasks_from_operator(left, tasks, task_by_op)) {
+    if (auto left_subtree_root = add_operator_tasks_recursively(left, tasks)) {
       left_subtree_root->set_as_predecessor_of(task);
     }
   }
   if (auto right = op->mutable_right_input()) {
-    if (auto right_subtree_root = _add_tasks_from_operator(right, tasks, task_by_op)) {
+    if (auto right_subtree_root = add_operator_tasks_recursively(right, tasks)) {
       right_subtree_root->set_as_predecessor_of(task);
     }
   }
 
-  // Add AFTER the inputs to establish a task order where predecessor get executed before successors
-  tasks.push_back(task);
-
   return task;
 }
 
+}  // namespace
+
+namespace opossum {
+OperatorTask::OperatorTask(std::shared_ptr<AbstractOperator> op, SchedulePriority priority, bool stealable)
+    : AbstractTask(priority, stealable), _op(std::move(op)) {}
+
+std::string OperatorTask::description() const {
+  return "OperatorTask with id: " + std::to_string(id()) + " for op: " + _op->description();
+}
+
+std::pair<std::vector<std::shared_ptr<AbstractTask>>, std::shared_ptr<OperatorTask>>
+OperatorTask::make_tasks_from_operator(const std::shared_ptr<AbstractOperator>& op) {
+  std::unordered_set<std::shared_ptr<OperatorTask>> operator_tasks_set;
+  const auto& root_operator_task = add_operator_tasks_recursively(op, operator_tasks_set);
+
+  return std::make_pair(
+      std::vector<std::shared_ptr<AbstractTask>>(operator_tasks_set.begin(), operator_tasks_set.end()),
+      root_operator_task);
+}
+
 const std::shared_ptr<AbstractOperator>& OperatorTask::get_operator() const { return _op; }
 
+void OperatorTask::skip_operator_task() {
+  Assert(shared_from_this().use_count() == 2, "Expected this OperatorTask to have a single owner.");
```

Comment:
I added 
```cpp
  // Newly created tasks always have TaskState::Created. However, AbstractOperator::get_or_create_operator_task needs
  // to create an OperatorTask in TaskState::Done, if the operator has already executed. This function provides a quick
  // path to TaskState::Done using dummy transactions.
  // While performing dummy transactions, a task should not be used elsewhere, e.g., in the scheduler, to prevent
  // race conditions, conflicts etc. Thus, the following Assert checks the use_count of the task's shared pointer.
  // Only the owning instance should hold a shared pointer to the task. But, since this function needs to create
  // another shared pointer to query the use_count, the Assert checks for `use_count() == 2` as follows:
```